File: src/ssnp_model.cpp
#include "ssnp_model.h"
#include <cmath>
#include <vector>
#include <complex>
#include <cassert>
#include <iostream>

std::vector<float> scatter_factor(const std::vector<float>& n, float res_z, float dz, float n0) {
    float factor = std::pow(2 * M_PI * res_z / n0, 2) * dz;
    std::vector<float> result(n.size());
    for (size_t i = 0; i < n.size(); ++i) {
        result[i] = factor * n[i] * (2 * n0 + n[i]);
    }
    return result;
}

std::vector<float> near_0(int size) {
    std::vector<float> result(size);
    for (int i = 0; i < size; ++i) {
        result[i] = std::fmod((i / float(size)) + 0.5f, 1.0f) - 0.5f;
    }
    return result;
}

std::vector<std::vector<std::complex<float>>> c_gamma(const std::vector<float>& res, const std::vector<int>& shape) {
    // Generate c_beta and c_alpha
    int size_alpha = shape[0];
    int size_beta = shape[1];
    std::vector<float> c_alpha = near_0(size_alpha);
    std::vector<float> c_beta = near_0(size_beta);

    // Normalize by resolution
    for (int i = 0; i < size_alpha; ++i) {
        c_alpha[i] /= res[1];
    }

    for (int i = 0; i < size_beta; ++i) {
        c_beta[i] /= res[0];
    }

    std::vector<std::vector<std::complex<float>>> result(1, std::vector<std::complex<float>>(size_alpha * size_beta));
    float eps = 1E-8f; // epsilon to avoid sqrt of negative numbers
    for (int i = 0; i < size_alpha; ++i) {
        for (int j = 0; j < size_beta; ++j) {
            float val = 1.0f - (std::pow(c_alpha[i], 2) + std::pow(c_beta[j], 2));
            result[0][i * size_beta + j] = std::sqrt(std::max(val, eps));
        }
    }
    // Unsqueeze
    return {result};
}

std::pair<std::vector<std::vector<std::vector<std::complex<double>>>>, std::vector<std::vector<std::vector<std::complex<double>>>>>
diffract(
    const std::vector<std::vector<std::complex<double>>>& uf,
    const std::vector<std::vector<std::complex<double>>>& ub,
    const std::vector<float>& res, 
    float dz) {
    assert(uf.size() == ub.size() && uf[0].size() == ub[0].size());

    int batch_size = uf.size();
    int size = uf[0].size();  // Flattened total size

    // Calculate shape dimensions (height x width)
    auto shape = std::vector<int>{static_cast<int>(std::sqrt(size)), static_cast<int>(std::sqrt(size))};
    while (shape[0] * shape[1] != size) {
        shape[1] += 1;
        if (shape[0] * shape[1] > size) {
            shape[0] += 1;
            shape[1] = size / shape[0];
        }
    }
    int height = shape[0];
    int width = shape[1];

    // Compute cgamma and kz
    auto cgamma = c_gamma(res, shape);
    std::vector<std::vector<std::complex<double>>> kz(batch_size, std::vector<std::complex<double>>(size));
    std::vector<std::vector<std::complex<double>>> eva(batch_size, std::vector<std::complex<double>>(size));

    double pi = 3.141592653589793;
    for (int b = 0; b < batch_size; ++b) {
        for (size_t i = 0; i < size; ++i) {
            kz[b][i] = std::complex<float>(2.0f * pi * res[2]) * cgamma[0][i];
            eva[b][i] = std::exp(std::min(std::abs((cgamma[0][i] - std::complex<float>(0.2)) * 5.0f), 0.0f));
        }
    }

    // Calculate p_mat
    std::vector<std::vector<std::vector<std::complex<double>>>> p_mat(4,
        std::vector<std::vector<std::complex<double>>>(batch_size, 
            std::vector<std::complex<double>>(size)));

    for (int b = 0; b < batch_size; ++b) {
        for (size_t i = 0; i < size; ++i) {
            std::complex<double> kz_val = kz[b][i];
            p_mat[0][b][i] = std::cos(kz_val * std::complex<double>(dz));
            p_mat[1][b][i] = std::sin(kz_val * std::complex<double>(dz)) / (kz_val + 1e-8);
            p_mat[2][b][i] = -std::sin(kz_val * std::complex<double>(dz)) * (kz_val);
            p_mat[3][b][i] = std::cos(kz_val * std::complex<double>(dz));

            for (int j = 0; j < 4; ++j) {
                p_mat[j][b][i] *= eva[b][i];
            }
        }
    }

    // Initialize reshaped uf_new and ub_new
    std::vector<std::vector<std::vector<std::complex<double>>>> uf_new(batch_size,
        std::vector<std::vector<std::complex<double>>>(height, std::vector<std::complex<double>>(width)));

    std::vector<std::vector<std::vector<std::complex<double>>>> ub_new(batch_size,
        std::vector<std::vector<std::complex<double>>>(height, std::vector<std::complex<double>>(width)));

    // Compute uf_new and ub_new and reshape
    for (int b = 0; b < batch_size; ++b) {
        for (size_t i = 0; i < size; ++i) {
            int row = i / width;
            int col = i % width;

            uf_new[b][row][col] = p_mat[0][b][i] * uf[b][i] + p_mat[1][b][i] * ub[b][i];
            ub_new[b][row][col] = p_mat[2][b][i] * uf[b][i] + p_mat[3][b][i] * ub[b][i];
        }
    }

    return {uf_new, ub_new};
}

std::vector<std::vector<bool>> binary_pupil(
    const std::vector<int>& shape, 
    float na, 
    const std::vector<float>& res) {
    auto cgamma = c_gamma(res, shape);
    size_t height = shape[0];
    size_t width = shape[1];

    std::vector<std::vector<bool>> mask(height, std::vector<bool>(width));
    double threshold = std::sqrt(1 - na * na);
    const auto& cgamma_batch = cgamma[0];

    // Compare each value of cgamma with the threshold and assign to mask
    for (size_t i = 0; i < height; ++i) {
        for (size_t j = 0; j < width; ++j) {
            // Use std::abs to compute the magnitude of the complex number
            mask[i][j] = std::abs(cgamma_batch[i * width + j]) > threshold;
        }
    }

    return mask;
}

std::vector<std::vector<std::vector<std::complex<double>>>> tilt(
    const std::vector<int>& shape,
    const std::vector<double>& angles,
    double NA,
    const std::vector<double>& res,
    bool trunc) {
    std::vector<std::vector<double>> c_ba(shape[0], std::vector<double>(2));

    // Compute c_ba (sine and cosine components)
    for (size_t i = 0; i < angles.size(); ++i) {
        c_ba[i][0] = NA * std::sin(angles[i]);
        c_ba[i][1] = NA * std::cos(angles[i]);
    }
    std::cout << "c_ba (sine and cosine components):\n";
    for (size_t i = 0; i < angles.size(); ++i) {
        std::cout << "(" << c_ba[i][0] << ", " << c_ba[i][1] << ")\n";
    }

    // Compute norm (shape * resolution)
    std::vector<double> norm = {shape[1] * res[1], shape[0] * res[0]};
    std::cout << "norm (shape * resolution):\n";
    std::cout << "(" << norm[0] << ", " << norm[1] << ")\n";

    // Compute factor (after truncation check)
    std::vector<std::vector<double>> factor(angles.size(), std::vector<double>(2));

    for (size_t i = 0; i < angles.size(); ++i) {
        for (size_t j = 0; j < 2; ++j) {
            double value = c_ba[i][j] * norm[j];
            if (trunc) {
                // Apply truncation if trunc is true
                factor[i][j] = std::trunc(value);  // Truncate the value
            } else {
                // Otherwise, store the non-truncated value
                factor[i][j] = value;
            }
        }
    }

    std::cout << "factor (after truncation check):\n";
    for (size_t i = 0; i < angles.size(); ++i) {
        std::cout << "(" << factor[i][0] << ", " << factor[i][1] << ")\n";
    }
}

File: tests/python/ssnp_model.py
"""SSNP forward model."""
import math
import torch
from torch import Tensor
from torch.nn import Module
from torch.types import Device

def scatter_factor(n: Tensor, res_z: float = 0.1, dz: float = 1, n0: float = 1.33) -> Tensor:
	return (2*torch.pi*res_z / n0)**2 * dz * n * (2*n0 + n)

def diffract(uf: Tensor, ub: Tensor, res: tuple[float, ...] = (0.1, 0.1, 0.1), dz: float = 1) -> tuple[Tensor, Tensor]:

	assert uf.shape == ub.shape, 'uf and ub must have the same shape'

	cgamma = c_gamma(res, uf.shape, device=uf.device)
	kz = 2 * torch.pi * res[0] * cgamma
	eva = torch.exp(torch.clamp((cgamma - 0.2) * 5, max=0))

	p_mat = torch.stack([torch.cos(kz * dz), torch.sin(kz * dz) / kz,
			-torch.sin(kz * dz) * kz, torch.cos(kz * dz)])
	
	p_mat *= eva

	uf_new1 = p_mat[0] * uf
	uf_new2 = p_mat[1] * ub
	uf_new = uf_new1 + uf_new2

	ub_new1 = p_mat[2] * uf
	ub_new2 = p_mat[3] * ub
	ub_new = ub_new1 + ub_new2

	return uf_new, ub_new

def c_gamma(res: tuple[float, ...], shape: tuple[int, ...], device: str | Device = 'cpu', dtype = torch.float32) -> Tensor:
	def _near_0(size):
		return torch.fmod(torch.arange(size, device=device).to(dtype) / size + 0.5, 1) - 0.5

	eps = 1E-8
	c_beta, c_alpha = [_near_0(size) / resolution for size, resolution in zip(shape[-2:], res[-2:])]
	return torch.sqrt(torch.clamp(1 - (torch.square(c_alpha) + torch.square(c_beta[:, None])), min=eps)).unsqueeze(0)

def binary_pupil(shape: tuple[int, ...], na: float, res: tuple[float, ...] = (0.1, 0.1, 0.1), device: str | Device = 'cpu') -> Tensor:
	cgamma = c_gamma(res, shape, device=device)
	mask = torch.greater(cgamma, (1 - na ** 2)**0.5)
	return mask

def tilt(shape: tuple[int, ...], c_ba, res, trunc: bool = True, device: str | Device = 'cpu') -> Tensor:

		norm = torch.tensor(shape) * torch.tensor(res[1:])
		norm = norm.view(1, 2)
		norm = norm.to(device=device)

		if trunc:
			factor = torch.trunc(c_ba * norm).T
		else:
			factor = (c_ba * norm).T

		xr = torch.arange(shape[1], device=device).view(1,1,-1).to(dtype=torch.complex128)
		xr = (2j * torch.pi / shape[1]) * factor[1].reshape(-1,1,1) * xr
		xr.exp_()

		yr = torch.arange(shape[0], device=device).view(1,-1,1).to(dtype=torch.complex128)
		yr = (2j * torch.pi / shape[0]) * factor[0].reshape(-1,1,1) * yr
		yr.exp_()

		out = xr * yr

		# normalize by center point value
		out /= out[:, *(i // 2 for i in shape)].clone().view(-1, 1, 1)
		return out

def merge_prop(uf: Tensor, ub: Tensor, res: tuple[float, ...] = (0.1, 0.1, 0.1)) -> tuple[Tensor, Tensor]:

	assert uf.device == ub.device, 'uf and ub must be on the same device'
	
	kz = c_gamma(res, uf.shape, device=uf.device) * (2 * torch.pi * res[0])

	uf_new = uf + ub
	ub_new = (uf - ub) * 1j * kz
	return uf_new, ub_new

def split_prop(uf: Tensor, ub: Tensor, res: tuple[float, ...] = (0.1, 0.1, 0.1)) -> tuple[Tensor, Tensor]:

	assert uf.device == ub.device, 'uf and ub must be on the same device'

	kz = c_gamma(res, uf.shape, device=uf.device) * (2 * torch.pi * res[0])

	ub_new = (uf + 1j*ub / kz) / 2
	uf_new = uf - ub_new

	return uf_new, ub_new

class SSNPBeam(Module):

	def __init__(
			self,
			res: tuple[float, ...] = (0.1, 0.1, 0.1),
			na: float = 0.65,
			angles: int = 32,
			intensity: bool = True,
		):
		super().__init__()

		self.res = res
		self.na = na
		self.angles = torch.nn.Parameter(torch.zeros(angles, 2), requires_grad=True)
		self.intensity = intensity

	def forward(self, n: Tensor) -> Tensor:

		shape = n.shape[-2:]
		
		# configure input feild
		Forward = torch.fft.fft2(tilt(shape, c_ba=self.angles, res=self.res, device=n.device))
		Backward = torch.zeros_like(Forward)
		U, UD = merge_prop(Forward, Backward, res=self.res)

	
		# propagate the wave through the RI distribution
		for slice in n:

			# propagate the wave 1.0*Δz
			U, UD = diffract(U, UD, res=self.res, dz=1.0)

			# convert feild to spatial domain
			u = torch.fft.ifft2(U)

			# compute scattering affects
			UD = UD - torch.fft.fft2(scatter_factor(slice, res_z=self.res[0], dz=1, n0=1.33) * u)

		# propagate the wave back to the focal plane
		U, UD = diffract(U, UD, res=self.res, dz = -len(n)/2)

		# merge the forward and backward feilds from u and ∂u
		Forward, _ = split_prop(U, UD, res=self.res)
		Forward = Forward * binary_pupil(shape, self.na, res=self.res, device=n.device)

		# return the intensity
		result = torch.abs(torch.fft.ifft2(Forward))
		return result**2 if self.intensity else result
